java-checks/src/main/java/org/sonar/java/checks/tests/AssertionCompareToSelfCheck.java

```diff
@@ -0,0 +1,189 @@
+    return MethodMatchers.or(
+      ASSERTJ_AND_FEST_ASSERT_SUBJECT_METHODS,
+      JUNIT5_ASSERTIONS,
+      JUNIT4_ASSERTIONS_WITH_MESSAGE,
+      JUNIT4_ASSERTIONS_WITHOUT_MESSAGE);
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree mit) {
+    if (JUNIT4_ASSERTIONS_WITH_MESSAGE.matches(mit)) {
+      checkActualAndExpectedExpression(methodName(mit).name(), mit.arguments().get(2), mit.arguments().get(1));
+    } else if (JUNIT4_ASSERTIONS_WITHOUT_MESSAGE.matches(mit) || JUNIT5_ASSERTIONS.matches(mit)) {
+      checkActualAndExpectedExpression(methodName(mit).name(), mit.arguments().get(1), mit.arguments().get(0));
+    } else {
+      // AssertJ or FestAssert
+      ExpressionTree actualExpression = mit.arguments().get(0);
+      consecutiveMethodIgnoringMessageDescription(mit)
+        .filter(ASSERTJ_AND_FEST_ASSERT_PREDICATES::matches)
+        .ifPresent(predicate -> checkActualAndExpectedExpression(methodName(predicate).name(), actualExpression, predicate.arguments().get(0)));
+    }
+  }
+
+  private void checkActualAndExpectedExpression(String predicateMethodName, ExpressionTree actualExpression, ExpressionTree expectedExpression) {
+    if (ExpressionsHelper.alwaysReturnSameValue(actualExpression) &&
+      SyntacticEquivalence.areEquivalent(actualExpression, expectedExpression) &&
+      !isLegitimateSelfComparison(predicateMethodName, actualExpression)) {
+      List<Location> secondaryLocations = Collections.singletonList(new Location("actual", actualExpression));
+      reportIssue(expectedExpression, MESSAGE, secondaryLocations, null);
+    }
+  }
+
+  private static boolean isLegitimateSelfComparison(String comparisonMethodName, ExpressionTree actualExpression) {
+    // In a unit test validating "equals" and "hashCode" methods, it's legitimate to compare an object to itself.
+    // In other kinds of tests, it's a bug. But it's complicated to know if a unit test is about validating the
+    // "equals" method or not. The following trade-off allows the self-comparison of an object if the unit test
+    // package is related to the tested object package.
+    Type actualExpressionType = actualExpression.symbolType();
+    return EQUALS_HASH_CODE_METHODS.contains(comparisonMethodName) &&
+      !isPrimitiveOrNull(actualExpressionType) &&
+      arePackageRelated(
+        packageOf(actualExpressionType.symbol()),
+        packageOf(classSymbolOfTheUnitTest(actualExpression)));
+  }
+
+  private static boolean isPrimitiveOrNull(Type actualExpressionType) {
+    return actualExpressionType.isPrimitive() || "null".equals(actualExpressionType.symbol().name());
+  }
+
+  static boolean arePackageRelated(String actualTypePackage, String unitTestPackage) {
+    return unitTestPackage.startsWith(actualTypePackage) || actualTypePackage.startsWith(unitTestPackage);
+  }
+
+  @CheckForNull
+  static Symbol classSymbolOfTheUnitTest(@Nullable ExpressionTree actualExpression) {
```

Comment:
Can't we use `JUtils.enclosingClass` here?

Reply:
:metal: